pandas/core/arrays/masked.py

```diff
@@ -217,6 +218,18 @@ def copy(self: BaseMaskedArrayT) -> BaseMaskedArrayT:
         mask = mask.copy()
         return type(self)(data, mask, copy=False)
 
+    @doc(ExtensionArray.factorize)
+    def factorize(self, na_sentinel: int = -1) -> Tuple[np.ndarray, ExtensionArray]:
+        arr = self._data
+        mask = self._mask
+
+        codes, uniques = _factorize_array(arr, na_sentinel=na_sentinel, mask=mask)
+
+        # the hashtables don't handle all different types of bits
+        uniques = uniques.astype(self.dtype.numpy_dtype, copy=False)  # type: ignore
```

Comment:
I would prefer not to type: ignore this; I find the current hierarchy of things and the implicit requirements a little wonky

Reply:
I am happy to do a PR that tries to rework the class hierarchy, but can that be done separately? Because it is really unrelated to the rest of this PR.

Reply:
I think it *is* related - this PR introduces an implicit requirement in factorize that `self.dtype` has a `numpy_dtype` attribute which. I think mypy is being smart to flag that; would prefer again to not ignore the advice

Reply:
OK, I added a minimal BaseMaskedDtype class to ensure typing is correct. Can you take a look?

There is no other way to indicate the types to mypy than adding yet another abstract property as I did? (the base ExtensionArray class also has an abstract `dtype` property, typed as `ExtensionDtype`, so I am override an abstract property with a new abstract property, just to have the different return type annotation)

Reply:
I think mypy is right about this one. It only knows (or knew prior to your commit) that BaseMasedArray.dtype is ExtensionDtype, which doesn't necessarily have a `.numpy_dtype` attribute.

Reply:
Yes, I understand. But I am just wondering if there is no other way to indicate this type change to mypy instead of adding another (for the rest useless) abstract property

Eg like 

```
dtype: BaseMaskedDtype
```

Reply:
That might actually work, testing that

Reply:
> instead of adding another (for the rest useless)

+1 for limiting how much contortion we do to keep mypy happy

Reply:
So now we get:

pandas/core/arrays/boolean.py:282: error: Signature of "dtype" incompatible with supertype "BaseMaskedArray"

because BooleanArray types `dtype` to return "BooleanDtype", while BaseMaskedArray types it to be a "BaseMaskedDtype" (and BooleanDtype is a subclass of BaseMaskedDtype).

So that doesn't appear to work either?

Reply:
Any further idea how to fix the mypy complaint?

Reply:
sorry, late to the party here.

```diff
diff --git a/pandas/core/arrays/base.py b/pandas/core/arrays/base.py
index 7447d593a..917640220 100644
--- a/pandas/core/arrays/base.py
+++ b/pandas/core/arrays/base.py
@@ -7,7 +7,17 @@ An interface for extending pandas with custom arrays.
    without warning.
 """
 import operator
-from typing import Any, Callable, Dict, Optional, Sequence, Tuple, Union
+from typing import (
+    Any,
+    Callable,
+    Dict,
+    Generic,
+    Optional,
+    Sequence,
+    Tuple,
+    TypeVar,
+    Union,
+)
 
 import numpy as np
 
@@ -32,8 +42,10 @@ from pandas.core.sorting import nargsort
 
 _extension_array_shared_docs: Dict[str, str] = dict()
 
+ExtensionDtypeT = TypeVar("ExtensionDtypeT", bound=ExtensionDtype)
 
-class ExtensionArray:
+
+class ExtensionArray(Generic[ExtensionDtypeT]):
     """
     Abstract base class for custom 1-D array types.
 
@@ -373,7 +385,7 @@ class ExtensionArray:
     # ------------------------------------------------------------------------
 
     @property
-    def dtype(self) -> ExtensionDtype:
+    def dtype(self) -> ExtensionDtypeT:
         """
         An instance of 'ExtensionDtype'.
         """
diff --git a/pandas/core/arrays/masked.py b/pandas/core/arrays/masked.py
index 59adbb16b..6c2adfcad 100644
--- a/pandas/core/arrays/masked.py
+++ b/pandas/core/arrays/masked.py
@@ -34,7 +34,7 @@ class BaseMaskedDtype(ExtensionDtype):
         raise AbstractMethodError
 
 
-class BaseMaskedArray(ExtensionArray, ExtensionOpsMixin):
+class BaseMaskedArray(ExtensionArray[BaseMaskedDtype], ExtensionOpsMixin):
     """
     Base class for masked arrays (which use _data and _mask to store the data).
 
@@ -43,7 +43,6 @@ class BaseMaskedArray(ExtensionArray, ExtensionOpsMixin):
 
     # The value used to fill '_data' to avoid upcasting
     _internal_fill_value: Scalar
-    dtype: BaseMaskedDtype
 
     def __init__(self, values: np.ndarray, mask: np.ndarray, copy: bool = False):
         # values is supposed to already be validated in the subclass
```

Reply:
@simonjayhawkins thanks for looking into it!

I find this quite complex though .. Also, wondering, what is the impact of adding such Generic to the actual class inheritance? In the python docs, as far as I can see, `Generic` is only used to create user defined type variables that then can be used in annotations, but not for subclassing actual classes.

Reply:
> Also, wondering, what is the impact of adding such Generic to the actual class inheritance?

The Generic base class uses a metaclass that defines __getitem__.  types are erased at runtime. 

Using generic classes (parameterized or not) to access attributes will result in type check failure. Outside the class definition body, a class attribute cannot be assigned, and can only be looked up by accessing it through a class instance that does not have an instance attribute with the same name

I believe the above restrictions applies to 3.6 as Generic was changed in 3.7 to not use a custom metaclass. see https://docs.python.org/3/library/typing.html#user-defined-generic-types

> In the python docs, as far as I can see, `Generic` is only used to create user defined type variables that then can be used in annotations, but not for subclassing actual classes.

see https://www.python.org/dev/peps/pep-0484/#instantiating-generic-classes-and-type-erasure

Reply:
@simonjayhawkins since I don't fully grasp the generic, yet, I updated it with a simpler abstract dtype property, which also seems to fix it. Or you OK with this, for now?

Reply:
> @simonjayhawkins since I don't fully grasp the generic, yet, I updated it with a simpler abstract dtype property (see the last added commit), which also seems to fix it. Or you OK with this, for now?

I'll look in a short while.

I know there is significant resistance to Generic, but it would allow say ExtensionArray[PeriodDtype] to be equivalent to PeriodArray and then AnyArrayLike could then be parameterised to say AnyArrayLike[PeriodDtype]. I suspect at some point we will want to refine types. but that's a discussion for another day.

Reply:
I am certainly open to going that way if gives better typing in general for EAs and solves a bunch of things. But maybe that can then go into a separate PR (at least I am not doing a `# type: ignore` anymore ;-))

Reply:
> Or you OK with this, for now?

LGTM

> But maybe that can then go into a separate PR

no problem. Generic (and Protocol) are the two typing additions that actually could potentially alter runtime behaviour, so it is right that we scrutinize this. The only problem i've encountered is the repr of type(\<instance\>) in an assert statement in py3.6, where the instance is an instance of a Generic Class. xref #31574

> at least I am not doing a `# type: ignore` anymore ;-)

👍